gravity.sh

```diff
@@ -597,12 +608,37 @@ gravity_DownloadBlocklistFromUrl() {
   mv "${listCurlBuffer}" "${listCurlBuffer%.*}.phgpb"
   listCurlBuffer="${listCurlBuffer%.*}.phgpb"
 
-  # Determine if $saveLocation has read permission
-  if [[ -r "${saveLocation}" && $url != "file"* ]]; then
-    # Have curl determine if a remote file has been modified since last retrieval
-    # Uses "Last-Modified" header, which certain web servers do not provide (e.g: raw github urls)
-    # Note: Don't do this for local files, always download them
-    heisenbergCompensator="-z ${saveLocation}"
+  # For all remote files, we try to determine if the file has changed to skip
+  # downloading them whenever possible.
+  if [[ $url != "file"* ]]; then
+    # Use the HTTP ETag header to determine if the file has changed if supported
+    # by curl. Using ETags is supported by raw.githubusercontent.com URLs.
+    if [[ "${etag_support}" == true ]]; then
+      # Save HTTP ETag to the specified file. An ETag is a caching related header,
+      # usually returned in a response. If no ETag is sent by the server, an empty
+      # file is created and can later be used consistently.
+      modifiedOptions="--etag-save ${saveLocation}.etag"
+
+      if [[ -f "${saveLocation}.etag" ]]; then
+        # This option makes a conditional HTTP request for the specific ETag read
+        # from the given file by sending a custom If-None-Match header using the
+        # stored ETag. This way, the server will only send the file if it has
+        # changed since the last request.
+        modifiedOptions="${modifiedOptions} --etag-compare ${saveLocation}.etag"
+      fi
+    fi
```

Comment:
Maybe we should use an `else` here. I was looking for how `etag` and `If-Modified-Since` relate to each other, but only found examples using one or the other:

https://curl.se/mail/archive-2022-03/0010.html
https://everything.curl.dev/http/modify/conditionals.html
https://daniel.haxx.se/blog/2019/12/06/curl-speaks-etag/

(and within the `-z` we probably also want `--remote-time`)

Reply:
I checked the code in `curl` and we should definitely use both at the same time. `-z` sets the time in the `If-Modified-Since` header and the others control the ETags. By setting both, we assure the best compatibility with any servers users may be downloading stuff from.

And I also don't think we need `--remote-time`. If we download today (Sunday), we can ask next time "only download if never than Sunday". It doesn't matter if the remote file has last been actually modified weeks ago. For some protocols (like FTL) this will require an extra step to get the timestamp, also I'd find is quite artificial to have `curl` create a file with a timestamp that is maybe years ago. But I also have no hard feeling against if you still think this is a good idea after reading my comment.

Reply:
Actually, after having read on the links you provided, I dare to differ with the statement

> Specifying both is rarely useful - there are corner cases where it may help.

most exactly because of what they say two sentences earlier

> Etag only applies to http. At the server, Etags take precedence over last-modified.

We support more protocols than HTTP alone.

Reply:
My concern was only that the time checking algorithm could fail due to different time zones when we use the local time stamp.

Reply:
ETags should always be preferred by the server when it supports them. `-z` is really just a legacy thing. We've been using it for long and I don't recall it ever causing any trouble for us. We're overall improving with also bringing in ETags.